File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.py
class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        h = set()
        for i in range(0, len(s) - k + 1):
            h.add(s[i:i+k])
        
        return len(h) == 2 ** k

# bit manipulation, hash map, string
# time: O(nk)
# space: O(2^k)

File: 1461-check-if-a-string-contains-all-binary-codes-of-size-k/1461-check-if-a-string-contains-all-binary-codes-of-size-k.go
func hasAllCodes(s string, k int) bool {
    h := map[string]bool{}
    for i := 0; i < len(s) - k + 1; i++ {
        if !h[s[i:i+k]] {
            h[s[i:i+k]] = true
        }
    }

    binSubTot := 1
    for i := 0; i < k; i++ {
        binSubTot *= 2
    }

    return len(h) == binSubTot
}

// bit manipulation, hash map, string
// time: O(nk)
// space: O(2^k)